packages/vm/lib/buildBlock.ts

```diff
@@ -0,0 +1,243 @@
+    }
+  }
+
+  /**
+   * Throws if the block has already been built or reverted.
+   */
+  checkStatus() {
+    if (this.built) {
+      throw new Error('Block has already been built')
+    }
+    if (this.reverted) {
+      throw new Error('State has already been reverted')
+    }
+  }
+
+  /**
+   * Returns the cumulative gas used by the transactions added to the block.
+   */
+  gasUsed(): BN {
+    return this.transactionResults.reduce(
+      (accumulator: BN, result: RunTxResult) => accumulator.iadd(result.gasUsed),
+      new BN(0)
+    )
+  }
+
+  /**
+   * Calculates and returns the transactionsTrie for the block.
+   */
+  private async transactionsTrie() {
+    const trie = new Trie()
+    for (const [i, tx] of this.transactions.entries()) {
+      await trie.put(encode(i), tx.serialize())
+    }
+    return trie.root
+  }
+
+  /**
+   * Calculates and returns the logs bloom for the block.
+   */
+  private bloom() {
+    const bloom = new Bloom()
+    for (const txResult of this.transactionResults) {
+      // Combine blooms via bitwise OR
+      bloom.or(txResult.bloom)
+    }
+    return bloom.bitvector
+  }
+
+  /**
+   * Calculates and returns the receiptTrie for the block.
+   */
+  private async receiptTrie() {
+    const gasUsed = new BN(0)
+    const receiptTrie = new Trie()
+    for (const [i, txResult] of this.transactionResults.entries()) {
+      const tx = this.transactions[i]
+      gasUsed.iadd(txResult.gasUsed)
+      const { encodedReceipt } = await generateTxReceipt.bind(this.vm)(tx, txResult, gasUsed)
+      await receiptTrie.put(encode(i), encodedReceipt)
+    }
+    return receiptTrie.root
+  }
+
+  /**
+   * Rewards the miner.
+   */
+  private async rewardMiner() {
+    const minerReward = new BN(this.vm._common.param('pow', 'minerReward'))
+    const reward = calculateMinerReward(minerReward, 0)
+    const coinbase = this.headerData.coinbase
+      ? new Address(toBuffer(this.headerData.coinbase))
+      : Address.zero()
+    await rewardAccount(this.vm.stateManager, coinbase, reward)
+  }
+
+  /**
+   * Run and add a transaction to the block being built.
+   * Please note that this modifies the state of the VM.
+   * Throws if the transaction's gasLimit is greater than
+   * the remaining gas in the block.
+   */
+  async addTransaction(tx: TypedTransaction) {
+    this.checkStatus()
+
+    if (!this.checkpointed) {
+      await this.vm.stateManager.checkpoint()
+      this.checkpointed = true
+    }
+
+    // According to the Yellow Paper, a transaction's gas limit
+    // cannot be greater than the remaining gas in the block
+    const blockGasLimit = new BN(toBuffer(this.headerData.gasLimit))
+    const blockGasRemaining = blockGasLimit.sub(this.gasUsed())
+    if (tx.gasLimit.gt(blockGasRemaining)) {
+      throw new Error('tx has a higher gas limit than the remaining gas in the block')
+    }
+
+    const result = await this.vm.runTx({ tx })
```

Comment:
If I am not misreading the code I think what we are not taking into account here yet at all is this miner state update in [runTx()](https://github.com/ethereumjs/ethereumjs-monorepo/blob/master/packages/vm/lib/runTx.ts#L273) based on the coinbase account.

Here the `runTx()` call is not passing in a `Block` object (which is by design I guess), but then the `coinbase` used will be wrong. Not sure what's the solution here, maybe a dedicated `miner` (`coinbase`?) option in `runTx()` or - broader - a similar optional `headerData` (or: `blockHeaderData` for clarity) option as introduced here? 🤔

Reply:
ah right, thanks. I was wondering why my spec test didn't catch this as I was getting a consistent final stateRoot, but it's because the coinbase defaults to the zero address so if I set the buildBlock coinbase to a non-zero address then it would have thrown. I will see if I can construct a minimal block to pass in for runTx, not only for this, but in case other things are added to runTx that are dependent on the passed-in block.